### PR TITLE
Missing docblocks affecting code

### PR DESCRIPTION
**Pull request by gmponos**
Fixing some Docblocks and typehints that are "affecting" the code in a way.

I was inspired to have a look to the repo by [this PR](https://github.com/guzzle/guzzle/pull/2217)

I should have added these changes on that one but I didn't thought about it... 

I have found more "minor" stuff affecting the code but I am trying not to mix unrelated things together in the PRs.

### CODE REVIEW

File: src/Cookie/CookieJar.php [hunk 1]
```diff
@@ -22,7 +22,7 @@ class CookieJar implements CookieJarInterface
      *                           arrays that can be used with the SetCookie
      *                           constructor
      */
-    public function __construct($strictMode = false, $cookieArray = [])
+    public function __construct($strictMode = false, array $cookieArray = [])
```
Comment (Nyholm) [inline src/Cookie/CookieJar.php]: This is technically a BC break.. Or i mean. It has always been the intention but people will have problem if we merge this.

File: src/RedirectMiddleware.php [hunk 2]
```diff
@@ -172,7 +172,7 @@ public function modifyRequest(
         // would do.
         $statusCode = $response->getStatusCode();
         if ($statusCode == 303 ||
-            ($statusCode <= 302 && $request->getBody() && !$options['allow_redirects']['strict'])
```
Comment (Nyholm) [inline src/RedirectMiddleware.php]: Why was this removed?

File: src/TransferStats.php [hunk 3]
```diff
@@ -18,18 +18,18 @@
...
+     * @param mixed                  $handlerErrorData Handler error data.
+     * @param array                  $handlerStats     Handler specific stats.
      */
     public function __construct(
         RequestInterface $request,
         ResponseInterface $response = null,
         $transferTime = null,
         $handlerErrorData = null,
-        $handlerStats = []
+        array $handlerStats = []
```
Comment (Nyholm) [inline src/TransferStats.php]: This is fine because the class if final

File: src/Cookie/CookieJar.php (hunk 1 above)
Comment (gmponos) [reply src/Cookie/CookieJar.php]: What I had in mind:

1. Constructor is allowed to be overridden in PHP and it is not required to follow parent.
2. If some has passed a different value rather than array it would still have failed on foreach...

but now that I think of it.. in cases where someone was passing an `iterable` this will fail...  So yea I will revert. 

😢 👍

File: src/RedirectMiddleware.php (hunk 2 above)
Comment (gmponos) [reply src/RedirectMiddleware.php]: Am I wrong? :/

File: src/Cookie/CookieJar.php (hunk 1 above)
Comment (gmponos) [reply src/Cookie/CookieJar.php]: Reverted... resolving.

File: src/TransferStats.php (hunk 3 above)
Comment (gmponos) [reply src/TransferStats.php]: Reverted... resolving.

File: src/RedirectMiddleware.php (hunk 2 above)
Comment (Nyholm) [reply src/RedirectMiddleware.php]: Nope. You are correct https://github.com/guzzle/psr7/blob/master/src/MessageTrait.php#L126

File: src/Handler/CurlFactory.php [hunk 4]
```diff
@@ -161,7 +161,7 @@ private static function createRejection(EasyHandle $easy, array $ctx)
 
         // If an exception was encountered during the onHeaders event, then
         // return a rejected promise that wraps that exception.
-        if ($easy->onHeadersException) {
+        if ($easy->onHeadersException instanceof \Exception) {
```
Comment (gmponos) [inline src/Handler/CurlFactory.php]: In a way this is also a BC...

If someone passed on `onHeadersException` a `StreamInterface` it would fail on the next line.. but now it wont.. and you will silently not see the error.

Comment (gmponos) [reply src/Handler/CurlFactory.php]: Maybe I must revert this as well? WDYT?

Comment (Nyholm) [reply src/Handler/CurlFactory.php]: I hesitated on this line. Let’s be safe rather then sorry. 

Revert it.

File: src/TransferStats.php (hunk 3 above)
Comment (gmponos) [reply src/TransferStats.php]: @Nyholm ok.. l am unresolving this let's continue the discussion here

1. As I said even if it was not final it does not actually matter since I can do this:
```
public function __constructor(){
    parent::__constructor('myval');
}
```

One constructor does not have to follow the same signature as the parent constructor.

2. The strong argument here is not about final or not... The main point here is this:

```
$stats = new TransferStats($request, null, null, null, null, null)
$stats->getHandlerStats(); // someone is expecting null here... although the docblock says otherwise...
```

As you said.. better be safe than sorry.
